### PR TITLE
Improve training stability and refactor codebase for simplicity

### PR DESCRIPTION
**Pull request by slowmoyang**
## Summary

This PR restructures the project from `deepmuonreco` into `muonly`, simplifies the training/configuration stack around instantiated Hydra components, and fixes several issues in training, inference, metrics, and job submission. It also migrates the project tooling to `uv` and cleans up developer workflow support.

### Package and tooling

- Rename the Python package from `deepmuonreco` to `muonly`
- Add `pyproject.toml` and `uv.lock`
- Use `uv_build` as the build backend
- Add Ruff configuration, project/dev dependencies, `.python-version`, and a `justfile`
- Add repository workflow/documentation improvements, including commit prompt guidance

### Config and architecture

- Replace the older monolithic config/model path with explicit `model`, `net`, `data`, and `optim` config groups
- Introduce `config/tts.yaml` as the main training entrypoint
- Consolidate detector features, preprocessing, and dataset path configuration into the data config
- Remove obsolete or broken configs, including deprecated criteria/feature/preprocessing/postprocessing groups and `tts_run3.yaml`
- Replace the generic `eval` OmegaConf resolver with dedicated resolvers such as `len` to reduce risk and improve clarity
- Instantiate models directly from Hydra config instead of relying on the previous `Model.from_config` path

### Model, training, and inference correctness

- Add a new `muonly.models.Model` base class and `TrackerTrackSelectionModel`
- Rewrite optimizer parameter grouping and learning-rate scheduling around AdamW with linear warmup and cosine decay
- Fix `predict.py` imports and checkpoint-based model instantiation for the renamed package/config structure
- Fix `predict_step` to call `self.net(batch)` directly instead of going through unintended Lightning behavior
- Correct several model/training bugs, including incorrect variable references and ineffective scheduler wiring
- Switch the vanilla transformer to pre-norm mode (`norm_first=True`) for more stable optimization

### Data pipeline and preprocessing

- Rewrite the datamodule and dataset flow under `muonly.data`
- Extract and reuse a shared `Compose` utility for preprocessing transforms
- Apply preprocessing directly from config in the data pipeline
- Improve collate/padding behavior with `torch.nn.utils.rnn.pad_sequence`
- Add `pre_fit_validation` control for validation-set lifecycle management
- Fix teardown behavior to release validation data correctly and reduce unnecessary memory retention
- Improve data-loading logs, preprocessing instrumentation, and warnings for empty preprocessing pipelines

### Metrics, logging, and evaluation

- Add a `Histogram` metric with plotting support
- Fix histogram handling for variable-width bin edges
- Expand tracker-track-selection evaluation with loss, AUROC/ROC, and score distributions for muon vs background
- Update validation/test metric handling and Aim logging
- Apply CMS matplotlib styling globally and reduce noisy warnings

### Submission and workflow fixes

- Update `submit.py` to use the new default config name `tts`
- Derive the project root from script location instead of requiring `PROJECT_ROOT`
- Validate GPU and CPU counts before submission
- Switch submission code to `htcondor2`
- Increase default memory from `64GB` to `80GB`
- Print the submitted cluster ID for easier job tracking

## Why

The main goal is to make the training stack easier to reason about and safer to operate:

- clearer separation between model, network, data, and optimization concerns
- fewer legacy configuration paths and broken variants to maintain
- more reliable inference, evaluation, and submission behavior
- simpler packaging and local development with `uv`

## Reviewer Notes

This is a large refactor with functional fixes mixed in. The easiest review order is:

1. package/build migration and rename to `muonly`
2. config reorganization and Hydra instantiation changes
3. model/data pipeline updates
4. inference and submission fixes
5. metrics, plotting, and logging improvements

## Suggested Validation

Recommended lightweight validation:

```bash
python ./train.py debug=sanity-check
```

### CODE REVIEW

File: src/muonly/data/datamodule.py [hunk 1]
```diff
@@ -0,0 +1,199 @@
...
+                **kwargs,
+            )
+        _logger.info(
+            f"Loaded {prefix} set with {len(dataset)} examples in {elapsed_time():.2f} seconds"
+        )
+
+        _logger.info(f"Preprocessing {prefix} set")
+        preprocessing = self.hparams["preprocessing"]
+        dataset.apply(preprocessing)
+        _logger.info(f"Preprocessed {prefix} set in {elapsed_time():.2f} seconds")
```
[the remark this answers is not in the document]
Comment (slowmoyang) [reply src/muonly/data/datamodule.py]: fixed via [03ecf3b](https://github.com/nurion4hep/DeepMuonReco/pull/41/commits/03ecf3b227cc4bc45bc5001b187c6a7635481148)

File: src/muonly/metrics/histogram.py [hunk 2]
```diff
@@ -0,0 +1,152 @@
...
+
+        self.histogram += histogram.hist.to(self.histogram.device)
+
+    def compute(self) -> Tensor | aim.Distribution:
+        return self.histogram
+
+    @cached_property
+    def edges(self) -> np.ndarray:
+        return np.linspace(*self.range, num=(self.bins + 1))
+
```
[the remark this answers is not in the document]
Comment (slowmoyang) [reply src/muonly/metrics/histogram.py]: fixed via [ec04c0f](https://github.com/nurion4hep/DeepMuonReco/pull/41/commits/ec04c0f02d66aefdbc0b7f04a0cf9568629dd3fb)

File: setup.sh
[the remark this answers is not in the document]
Comment (slowmoyang) [reply setup.sh]: removed it via [20836c0](https://github.com/nurion4hep/DeepMuonReco/pull/41/commits/20836c074b150a2f89db708c06da0b23bbf9c56d)
